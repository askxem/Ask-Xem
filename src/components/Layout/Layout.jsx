import Nav from '../Nav/Nav.jsx'

export default function Layout({ children }) {
  return (
    <div>
      <Nav />
      {children}
    </div>
  )
}
