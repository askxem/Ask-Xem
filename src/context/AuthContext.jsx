import { createContext, useState, useContext, useMemo } from "react";
import { getUser, signInUser, signOutUser, signUpUser } from "../services/users.js";


const AuthContext = createContext();

function ProvideAuth({children}) {

    const currentUser = getUser();

    const [user, setUser] = useState(
        currentUser ? {...currentUser} : {}
    );

    async function signUp(email, password) {
        const newUser = await signUpUser(email, password);
        setUser(newUser);
    }

    async function signIn(email, password) {
        const newUser = await signInUser(email, password);
        setUser(newUser);
    }

    async function signOut() {
        await signOutUser();
        localStorage.clear();
        setUser({});
        history.push('/home');
    }

    const value = useMemo(() => ({ user, signUp, signIn, signOut}), [user]);

    return (
        <AuthContext.Provider value={value} >
            {children}
        </AuthContext.Provider>
    )
}

/**
 * 
 * @returns a user object, a signup, signin, and signout fn that hits the database.
 */
function useAuth(){
    const context = useContext(AuthContext);

    if (context === undefined) throw new Error('Auth context not accessible outside of the auth provider.')

    return context;
}


export { ProvideAuth, useAuth }