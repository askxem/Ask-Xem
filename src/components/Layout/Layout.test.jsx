import { render} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Layout from './Layout.jsx';
import { ProvideAuth } from '../../context/AuthContext.jsx';

jest.mock('../../context/AuthContext.jsx');

it('renders layout component', () => {
    const { container } = render(
        <ProvideAuth>
            <MemoryRouter>
                <Layout />
            </MemoryRouter>
        </ProvideAuth>
    )

    expect(container).toMatchSnapshot();
})

