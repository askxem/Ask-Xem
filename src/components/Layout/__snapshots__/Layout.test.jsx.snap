// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders layout component 1`] = `
<div>
  <div>
    <header
      style="opacity: 0; transform: translateY(-100vh) translateZ(0);"
    >
      <a
        href="/"
      >
        <h1>
          ask xem
        </h1>
      </a>
      <article>
        <a
          href="/login"
        >
          Login
        </a>
      </article>
    </header>
  </div>
</div>
`;
